cpp/src/Slice/Grammar.y

```diff
@@ -385,6 +385,75 @@ module_def
         $$ = nullptr;
     }
 }
+| ICE_MODULE ICE_SCOPED_IDENTIFIER
+{
+    auto ident = dynamic_pointer_cast<StringTok>($2);
+    ContainerPtr cont = currentUnit->currentContainer();
+
+    // Split the scoped-identifier token into separate module names.
+    vector<string> modules;
+    size_t startPos = 0;
+    size_t endPos;
+    while ((endPos = ident->v.find("::", startPos)) != string::npos)
+    {
+        modules.push_back(ident->v.substr(startPos, (endPos - startPos)));
+        startPos = endPos + 2;
+    }
+    modules.push_back(ident->v.substr(startPos));
+
+    // Create the nested modules.
+    for (size_t i = 0; i < modules.size(); i++)
+    {
+        const auto currentModuleName = modules[i];
+        ModulePtr module = cont->createModule(currentModuleName);
+        if (module)
+        {
+            cont->checkIntroduced(currentModuleName, module);
+            currentUnit->pushContainer(module);
+            $$ = cont = module;
+        }
+        else
+        {
+            // If an error occurs while creating one of the modules, we have to stop. But, to eagerly report as many
+            // errors as possible, we still 'create' any remaining modules, which will run _some_ validation on them.
+            for (size_t j = (i + 1); j < modules.size(); j++)
+            {
+                cont->createModule(modules[j]); // Dummy
+            }
+
+            // Then we roll back the chain, ie. pop the successfully-created-modules off the container stack.
+            for (; i > 0; i--)
+            {
+                currentUnit->popContainer();
+            }
+            $$ = nullptr;
+            break;
+        }
+    }
+}
+'{' definitions '}'
+{
+    if ($3)
+    {
+        // Set the 'return value' to the inner-most module (the current module, before we've popped any off).
+        // Whichever module we return, is the one that metadata will be applied to.
+        $$ = currentUnit->currentContainer();
+
+        // We need to pop '(N+1)' modules off the container stack, to navigate out of the nested module.
+        // Where `N` is the number of scope separators ("::").
+        size_t scopePos = 0;
+        auto ident = dynamic_pointer_cast<StringTok>($2);
+        while ((scopePos = ident->v.find("::", scopePos + 2)) != string::npos)
```

Comment:
Why the +2, does ident always start with `::`?

Reply:
The intent was to skip over `"::"` in between module identifiers...
But you're right, this won't work for something like `A::B`...

Reply:
Fixed.